### PR TITLE
Choices

### PR DESCRIPTION
**Pull request by dcol97**
This PR introduces two RamOperations, RamChoice and RamIndexChoice. These operations combine a RamScan or RamIndexScan and a RamFilter into a single operation, RamChoice or RamIndexChoice respectively, whereby the one element satisfying the RamFilter's condition is returned (if found).

These operations form the basis of the following transformation:
```
 QUERY
  ...
   SEARCH t1 IN A INDEX t1.x=10 AND t1.y = 20
    	IF (t1.x, t1.y) NOT IN A
        ... // no occurrence of t1

will be rewritten to

 QUERY
  ...
   INDEXCHOICE AS t1 ON INDEX t1.x=10 AND t1.y = 20
    WHERE (t1.x, t1.y) NOT IN A
    ... // no occurrence of t1
```
Since the rest of the loop nest does not depend on t1 (for example), the same computation may be performed multiple times without the transformation.

### CODE REVIEW

File: src/RamIndexKeys.cpp [hunk 1]
```diff
@@ -31,4 +31,16 @@ SearchColumns RamIndexKeysAnalysis::getRangeQueryColumns(const RamIndexRelationS
     return keys;
 }
 
+/** Get indexable columns of index choice*/
+SearchColumns RamIndexKeysAnalysis::getRangeQueryColumns(const RamIndexChoice* search) const {
```
Comment (b-scholz) [inline src/RamIndexKeys.cpp]: We don't need that method. This is subsumed by  ```RamIndexKeysAnalysis::getRangeQueryColumns(const RamIndexRelationSearch ...)```.

File: src/RamTransforms.cpp [hunk 2]
```diff
@@ -295,4 +295,125 @@ bool IfConversionTransformer::convertIndexScans(RamProgram& program) {
...
+                    }
+                } else if (const RamUnpackRecord* unpack = dynamic_cast<const RamUnpackRecord*>(&node)) {
+                    if (unpack->getReferenceLevel() == scan->getIdentifier()) {
+                        tupleNotUsed = false;
+                    }
+                }
+            });
+        }
+    } else {
+        tupleNotUsed = false;
```
[the remark this answers is not in the document]
Comment (dcol97) [reply src/RamTransforms.cpp]: Yep, can be done with one variable.

File: src/RamOperation.h [hunk 3]
```diff
@@ -288,6 +288,146 @@ class RamIndexScan : public RamIndexRelationSearch {
...
+    /** get condition */
+    const RamCondition& getCondition() const {
+        return *condition;
+    }
+
+    /** Print */
+    void print(std::ostream& os, int tabpos) const override {
+        const RamRelation& rel = getRelation();
+        os << times(" ", tabpos);
+        os << "CHOICE " << rel.getName() << " AS t" << getIdentifier() << " ON INDEX ";
```
[the remark this answers is not in the document]
Comment (dcol97) [reply src/RamOperation.h]: Sure.

File: src/RamIndexKeys.h [hunk 4]
```diff
@@ -34,6 +34,9 @@ class RamIndexKeysAnalysis : public RamAnalysis {
 
     /** Get indexable columns of index scan */
     SearchColumns getRangeQueryColumns(const RamIndexRelationSearch* search) const;
+
+    /** Get indexable columns of index choice */
+    SearchColumns getRangeQueryColumns(const RamIndexChoice* search) const;
```
[the remark this answers is not in the document]
Comment (dcol97) [reply src/RamIndexKeys.h]: Remnant from before RamIndexChoice was a subclass of RamIndexRelationSearch I guess :)

File: src/RamVisitor.h [hunk 5]
```diff
@@ -184,6 +186,8 @@ struct RamVisitor : public ram_visitor_tag {
     LINK(UnpackRecord, Search);
     LINK(Scan, RelationSearch);
     LINK(IndexScan, RelationSearch);
+    LINK(Choice, RelationSearch);
+    LINK(IndexChoice, RelationSearch);
```
[the remark this answers is not in the document]
Comment (dcol97) [reply src/RamVisitor.h]: > LINK(IndexChoice, IndexRelationSearch)

It doesn't compile when only this line is changed.

File: src/RamOperation.h [hunk 6]
```diff
@@ -288,6 +288,145 @@ class RamIndexScan : public RamIndexRelationSearch {
...
+            }
+        }
+        RamIndexChoice* res = new RamIndexChoice(std::unique_ptr<RamRelationReference>(relationRef->clone()),
+                getTupleId(), std::unique_ptr<RamCondition>(condition->clone()), std::move(resQueryPattern),
+                std::unique_ptr<RamOperation>(getOperation().clone()), getProfileText());
+        return res;
+    }
+
+protected:
+    /** Condition */
```
Comment (mmcgr) [inline src/RamOperation.h]: This is not a helpful comment. Either be verbose enough for it to describe the point of the condition, or remove it.

File: src/RamTransforms.cpp [hunk 7]
```diff
@@ -295,4 +295,115 @@ bool IfConversionTransformer::convertIndexScans(RamProgram& program) {
...
+
+    // Check that RamFilter follows the Scan in the loop nest
+    if (const auto* filter = dynamic_cast<const RamFilter*>(&scan->getOperation())) {
+        // Check that the Filter uses the identifier in the Scan
+        if (rcla->getLevel(&filter->getCondition()) == scan->getTupleId()) {
+            transformTuple = true;
+
+            // Check that the filter is not referred to after
+            const auto* nextNode = dynamic_cast<const RamNode*>(&filter->getOperation());
+            visitDepthFirst(*nextNode, [&](const RamNode& node) {
```
Comment (mmcgr) [inline src/RamTransforms.cpp]: This is an odd approach. Why not use two visitors, one for RamElementAccess and one for RamUnpackRecord?

File: src/RamTransforms.cpp [hunk 8]
```diff
@@ -295,4 +295,115 @@ bool IfConversionTransformer::convertIndexScans(RamProgram& program) {
...
+
+    // Check that RamFilter follows the IndexScan in the loop nest
+    if (const auto* filter = dynamic_cast<const RamFilter*>(&indexScan->getOperation())) {
+        // Check that the Filter uses the identifier in the IndexScan
+        if (rcla->getLevel(&filter->getCondition()) == indexScan->getTupleId()) {
+            transformTuple = true;
+
+            // Check that the filter is not referred to after
+            const auto* nextNode = dynamic_cast<const RamNode*>(&filter->getOperation());
+            visitDepthFirst(*nextNode, [&](const RamNode& node) {
```
Comment (mmcgr) [inline src/RamTransforms.cpp]: As above, odd approach, use two visitors.

File: src/Synthesiser.cpp [hunk 9]
```diff
@@ -651,6 +651,48 @@ void Synthesiser::emitCode(std::ostream& out, const RamStatement& stmt) {
...
+            });
+
+            // if nullary
+            if (choice.getRelation().isNullary()) {
+                out << "if(!" << relName << "->"
+                    << "empty()) {\n";
+                visitSearch(choice, out);
+                out << "}\n";
+            } else {
+                // TODO: parallel
```
Comment (mmcgr) [inline src/Synthesiser.cpp]:
```
                // TODO (usename/issue): Make run in parallel
```

File: src/LVMGenerator.h [hunk 10]
```diff
@@ -884,6 +968,18 @@ class LVMGenerator : protected RamVisitor<void, size_t> {
         return scanIteratorIndex++;
     }
 
+    /** Iter */
+    size_t choiceIteratorIndex = 0;
```
[the remark this answers is not in the document]
Comment (XiaowenHu96) [reply src/LVMGenerator.h]: Yes. Most of the legacy issues in LVM* files should be left to me. I'll start work on the refactoring of LVM in next week hopefully.

File: src/RamTransforms.cpp (hunk 7 above)
Comment (b-scholz) [reply src/RamTransforms.cpp]: A traversal is not cheap - and it is cheaper to check the cases inside a single visitor.

File: src/RamTransforms.cpp (hunk 8 above)
Comment (b-scholz) [reply src/RamTransforms.cpp]: Same consideration.

File: src/RamTransforms.cpp (hunk 7 above)
Comment (mmcgr) [reply src/RamTransforms.cpp]: With two visitors the purpose of the code is clearer, and the use of visitDepthFirst is more standard, thus quicker to read.

That said, the complexity of two traversals that do one operation is the same as one traversal that does two operations, surely? In this case we're performing 50% more dynamic casts if we combine the two vs double the iteration costs if they are separate, so any improvement in constant factors seems arguable. The RAM is also not large, so performance is not an issue here.

**Comment by b-scholz** [comment]
This is related to issue #163.

**Pull request merged**